cpp/ql/test/library-tests/dataflow/taint-tests/test_ir.expected

```diff
@@ -1,11 +1,13 @@
-| taint.cpp:8:8:8:13 | Load: clean1 | taint.cpp:4:27:4:33 | InitializeParameter: source1 |
-| taint.cpp:16:8:16:14 | Load: source1 | taint.cpp:12:22:12:27 | Call: call to source |
-| taint.cpp:17:8:17:16 | Add: ++ ... | taint.cpp:12:22:12:27 | Call: call to source |
-| taint.cpp:109:7:109:13 | Load: access to array | taint.cpp:105:12:105:17 | Call: call to source |
-| taint.cpp:129:7:129:9 | Load: * ... | taint.cpp:120:11:120:16 | Call: call to source |
-| taint.cpp:130:7:130:9 | Load: * ... | taint.cpp:127:8:127:13 | Call: call to source |
-| taint.cpp:134:7:134:9 | Load: * ... | taint.cpp:120:11:120:16 | Call: call to source |
-| taint.cpp:151:7:151:12 | Call: call to select | taint.cpp:151:20:151:25 | Call: call to source |
-| taint.cpp:167:8:167:13 | Call: call to source | taint.cpp:167:8:167:13 | Call: call to source |
-| taint.cpp:168:8:168:14 | Load: tainted | taint.cpp:164:19:164:24 | Call: call to source |
-| taint.cpp:210:7:210:7 | Load: x | taint.cpp:207:6:207:11 | Call: call to source |
+| taint.cpp:8:8:8:13 | clean1 | taint.cpp:4:27:4:33 | source1 |
+| taint.cpp:16:8:16:14 | source1 | taint.cpp:12:22:12:27 | ... = ... |
+| taint.cpp:16:8:16:14 | source1 | taint.cpp:12:22:12:27 | call to source |
+| taint.cpp:17:8:17:16 | ++ ... | taint.cpp:12:22:12:27 | ... = ... |
```

Comment:
Okay, I see what you're saying.

The rules for the return value of an assignment are different between C and C++. To keep this simple, let's focus on C++. Let me know if I've understood this correctly. A C++ assignment expression `a = b` has a result that's an lvalue for the same address as `a`. This is consistent with overloaded assignment operators, which look like `C& C::operator=(const C &rhs) { ...; return *this; }`. This makes it possible to implement a `class my_int { int content; ... }` for which assignment behaves exactly (okay, _exactly_ is a big word in this context) the same as the built-in assignment on `int`.

I'd argue that ideally the IR translation of a built-in `=` should work in the same way as an overloaded `operator=`. It'll have an lvalue as result, and then that lvalue will be `Load`ed if necessary (test case 1 and 3). And that seems to be how it works now.

So your suggestion is to add a `CopyValueInstruction` to `TranslatedAssignment`. For performance and clarity, we'd probably want to omit that instruction in the common case where `=` is in a void context.

I'm still not sure how common it is for `TranslatedElement`s to have results that are not their own instructions. If it's common, perhaps it will bite us in `DataFlow::asExpr` for other kinds of expression, and we should be looking for a way to fix all the cases in a unified way. Like automatically inserting `CopyValueInstruction` before a `TranslatedLoad` that would otherwise get its operand from a different AST element.

Reply:
Currently `TranslatedCommaExpr`, `TranslatedPrefixCrementOperation` in C++ when the result is not being loaded, `TranslatedTransparentExpr`, `TranslatedAssignment`, and `TranslatedConditionDeclExpr` have a `getResult` that is a child's `getResult`. `TranslatedDestructorFieldDestruction` has no `getResult`.

Of those, `TranslatedTransparentExpr` and `TranslatedAssignment` seem like the only ones that are likely to show up frequently as sources or sinks.

Reply:
Thanks for checking. It sounds to me like adding `CopyValueInstruction` to all these instructions would be a good solution. But that's not for this PR.

Do you think this PR should be delayed until we can improve the instruction-to-expr mapping? I think this PR is just exposing a problem that was there already.

Reply:
@rdmarsh2's analysis is spot on. While we've always tried to ensure that there's at least one unique `Instruction` for each leaf `Expr`, we've been willing to let interior `Expr` nodes return the result of a child. We could go back and add `CopyValue` in all of those cases that aren't already in a void context, but one consequence of that would be that every use of parentheses in an expression would get another `Instruction`, which seems extravagant. I think that targeting only those expressions that are likely to be used as sources or sinks would make more sense. Either that, or only target those expressions that are not off to the side as `Conversion`s. That may be the same set.

Reply:
I think that's essentially the same set.

Reply:
I've merged back from `master`, and this problem has gone away thanks to #1999.